File: 프로그래머스/프로그래머스_Lv2/우박수열 정적분/Solution.java
import java.util.*;

class Solution {
    
    private double[][] sum;
    
    public double[] solution(int k, int[][] ranges) {
        List<Double> areaSum = getAreaSum(getCollatz(k));
        sum = new double[areaSum.size()+1][areaSum.size()+1];
        for(int index = 0; index < areaSum.size(); index++) {
            sum[index][index+1] = areaSum.get(index);
        }
        double[] result = new double[ranges.length];
        for(int index = 0; index < ranges.length; index++) {
            int start = ranges[index][0];
            int end = areaSum.size() + ranges[index][1];
            result[index] = getAccumulatedSum(start, end);
        }
        return result;
    }
    
    public double getAccumulatedSum(final int start, final int end) {
        if (start == end) return 0;
        if(sum[start][end] != 0) return sum[start][end];
        int mid = (start+end)/2;
        sum[start][end] = getAccumulatedSum(start, mid) + getAccumulatedSum(mid, end);
        return sum[start][end];
    }
    
    public List<Double> getCollatz(int k) {
        List<Double> list = new ArrayList<>();
        double collatz = k;
        list.add(collatz);
        while(collatz != 1) {
            if(collatz% 2 == 0) {
                collatz /=2;
                list.add(collatz);
                continue;
            }
            collatz = 3 * collatz + 1;
            list.add(collatz);
        }
        return list;
    }
    
    public List<Double> getAreaSum(final List<Double> collatz) {
        List<Double> list = new ArrayList<>();
        for(int index = 1; index < collatz.size(); index++) {
            double size = (collatz.get(index) + collatz.get(index-1))/2;
            list.add(size);
        }
        return list;
    }
}

File: 프로그래머스/프로그래머스_Lv2/우박수열 정적분/solution.js
let sum;

function solution(k, ranges) {
  const arr = [];
  arr.push(k);
  while (k != 1) {
    k = parseValue(k);
    arr.push(k);
  }
  sum = Array.from({ length: arr.length }, () =>
    Array.from({ length: arr.length }, () => 0)
  );

  for (let index = 1; index < arr.length; index++) {
    sum[index - 1][index] = (arr[index - 1] + arr[index]) / 2;
  }
  return ranges.map((each) => {
    each[1] = arr.length + each[1] - 1;
    const min = Math.min(...each);
    const max = Math.max(...each);
    if (each[0] > each[1]) return -1;
    return calculatedAccumulated(min, max);
  });
}

function parseValue(k) {
  if (k % 2 === 0) {
    return k / 2;
  }
  return 3 * k + 1;
}

function calculatedAccumulated(start, end) {
  if (start === end) return 0;
  if (sum[start][end] !== 0) return sum[start][end];
  const middle = Math.floor((start + end) / 2);
  return (sum[start][end] =
    calculatedAccumulated(start, middle) + calculatedAccumulated(middle, end));
}
